app/models/providers/infra/vm_or_template.rb

```diff
@@ -32,6 +32,29 @@ class Infra::VmOrTemplate < ApplicationRecord
     has_one                   :hardware, :dependent => :destroy
     has_many                  :disks, :through => :hardware
 
+    def state
+      (power_state || "unknown").downcase
+    end
+    alias_method :current_state, :state
+
+    # Override raw_power_state= attribute setter in order to impose side effects
+    # of setting previous_state and updating state_changed_on
+    def raw_power_state=(new_state)
+      return unless new_state
+
+      unless raw_power_state == new_state
+        self.previous_state   = raw_power_state
+        self.state_changed_on = Time.now.utc
```

Comment:
this timestamp isn't really accurate right? it's really the time we do the refresh, right?

Reply:
All of these are copied&pasted from manageiq, https://github.com/ManageIQ/manageiq/blob/master/app/models/vm_or_template.rb#L1274

Reply:
and yeah technically its a few microseconds early since it hasn't been saved yet but it is good enough i guess haha